crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,551 @@
+
+    let mut available_authorities: BTreeSet<_> = _proposals.iter().map(|(auth, _)| *auth).collect();
+    available_authorities.remove(&_active_authority.state.name); // remove ourselves
+    let mut fragments_num = 0;
+
+    loop {
+        let next_checkpoint_sequence_number: u64 = checkpoint_db.lock().next_checkpoint();
+        if next_checkpoint_sequence_number > *_my_proposal.proposal.0.checkpoint.sequence_number() {
+            // Our work here is done, we have progressed past the checkpoint for which we were given a proposal.
+            // Our DB has been updated (presumably by consensus) with the sought information (a checkpoint
+            // for this sequence number)
+            break;
+        }
+
+        // We have ran out of authorities?
+        if available_authorities.is_empty() {
+            // We have created as many fragments as possible, so exit.
+            break;
+        }
+
+        let random_authority = _active_authority.state.committee.sample();
+        if available_authorities.remove(random_authority) {
+            // Get a client
+            let client = _active_authority.net.authority_clients[random_authority].clone();
+
+            if let Ok(response) = client
+                .handle_checkpoint(CheckpointRequest::latest(true))
+                .await
+            {
+                if let AuthorityCheckpointInfo::Proposal { current, previous } = &response.info {
+                    // Check if there is a latest checkpoint
+                    if let AuthenticatedCheckpoint::Certified(prev) = previous {
+                        if prev.checkpoint.sequence_number > next_checkpoint_sequence_number {
+                            // We are now way behind, return
+                            return;
+                        }
+                    }
+
+                    // For some reason the proposal is empty?
+                    if current.is_none() || response.detail.is_none() {
+                        continue;
+                    }
+
+                    // TODO: check the proposal is also for the same checkpoint sequence number?
+                    if current.as_ref().unwrap().0.checkpoint.sequence_number()
+                        != _my_proposal.sequence_number()
+                    {
+                        return;
+                    }
+
+                    let other_proposal = CheckpointProposal::new(
+                        current.as_ref().unwrap().clone(),
+                        response.detail.unwrap(),
+                    );
+
+                    let fragment = _my_proposal.fragment_with(&other_proposal);
+                    let _ = checkpoint_db.lock().handle_receive_fragment(&fragment);
+
+                    // TODO: here we should really wait until the fragment is sequenced, otherwise
+                    //       we would be going ahead and sequencing more fragments that may not be
+                    //       needed. For the moment we just linearly back-off.
```

Comment:
Do you want a feedback channel from consensus to notify you of when the fragment is sequenced?

Reply:
That would be very useful here indeed!

Reply:
It seems it is not easy without async. If the function `send_to_consensus` of the trait `ConsensusSender` is async, we can return a `oneshot::receiver` here. Should we leave it as a TODO for now?